Reworked lookup sends in LocatorDiscovery

bernardnormier opened the pull request:
This PR reworks the sending on lookup request in LocatorDiscovery. The lookups are now sent in parallel.

externl commented on csharp/src/Ice/LocatorDiscovery/Locator.cs:
```diff
@@ -271,45 +271,57 @@ internal void FoundLocator(ILocatorPrx locator)
+                    }
+                }
+
+                Task t = Task.WhenAll(invocationTasks);
```
Could this potentially hang if the user has specified a tcp lookup endpoint. I know they're not supposed to, but I'm not sure that we check. Is it otherwise possible for UDP to "hang" here?

bernardnormier replied:
The lookup proxy is created as a datagram proxy:
```
var lookupPrx = ILookupPrx.Parse($"IceLocatorDiscovery/Lookup -d:{lookupEndpoints}", _communicator);
lookupPrx = lookupPrx.Clone(clearRouter: false);
```

The user cannot override this code by specifying tcp endpoints.

We also have a:
```
 Debug.Assert(endpoint.Transport == Transport.UDP);
```
Which could be replaced by a throw InvalidConfigurationException is you're concerned about bogus configuration.

So it's definitely datagram/UDP. 

I guess the sending with UDP can take a little time. Since we're not setting an explicit invocation timeout, we're using the default invocation timeout, which is by default 60s. Maybe we should use the plugin's timeout (by default 300ms) as invocation timeout on these lookup proxies?

externl replied:
> Which could be replaced by a throw InvalidConfigurationException is you're concerned about bogus configuration.

Yes. I think it would indeed be better to throw an `InvalidConfigurationException` here instead.

externl replied:
> Maybe we should use the plugin's timeout (by default 300ms) as invocation timeout on these lookup proxies?

Sure, that's probably better than 60s

pepone commented:
Looks good, but wondering if would be a bit simpler to do something like

```
try
{
    await Task.WhenAll(_lookups.Select(entry =>
        {
            try
            {
                return entry.Key.FindLocatorAsync(_instanceName, entry.Value);
            }
            catch (Exception ex)
            {
                return Task.FromException(ex);
            }
        })).ConfigureAwait(false);
}
catch (AggregateException ex)
{
    if (ex.InnerExceptions.Count == _lookups.Count)
    {
```

bernardnormier commented:
> Looks good, but wondering if would be a bit simpler to do something like
> 

Yes, I can rework the code along those lines.

Note that `await Task.WhenAll(...)` does not throw an AggregateException, but the first inner exception of the aggregate exception.

bernardnormier commented:
> > Looks good, but wondering if would be a bit simpler to do something like
> 
> Yes, I can rework the code along those lines.
> 
> Note that `await Task.WhenAll(...)` does not throw an AggregateException, but the first inner exception of the aggregate exception.

Fixed as suggested. Also fixed IceDiscovery.

pepone commented:
Think a bit more about this and I wonder if we should just wait for one to success instead of waiting for all to finish, as soon as one task succeed we can start awaiting the completion source. 

this can be implemented using WhenAny, if the returned task succeeds we move on, and if it failed we call it again removing the returned task.

bernardnormier commented:
> Think a bit more about this and I wonder if we should just wait for one to success instead of waiting for all to finish, as soon as one task succeed we can start awaiting the completion source.
> 
> this can be implemented using WhenAny, if the returned task succeeds we move on, and if it failed we call it again removing the returned task.

See the latest code.

pepone reviewed (approved):
Latest code looks good

externl commented on csharp/src/Ice/Discovery/Locator.cs:
```diff
@@ -165,16 +166,17 @@ internal Locator(ILocatorRegistryPrx registry, ILookupPrx lookup, ObjectAdapter
-                Debug.Assert(endpoint.Transport == Transport.UDP);
+                if (!endpoint.IsDatagram)
+                {
+                    throw new InvalidConfigurationException("lookup can only have udp endpoints");
```
Should we reference the property here so the user isn't confused at to what "lookup" means?

bernardnormier replied:
Fixed.

externl reviewed (approved):
Looks good.
